sotodlib/hwp/hwp.py

```diff
@@ -6,429 +6,577 @@
 logger = logging.getLogger(__name__)
 
 
-def func_hwpss_2f4f(x, A2, A4, phi2, phi4):
-    return A2 * np.cos(2 * x + phi2) + A4 * np.cos(4 * x + phi4)
+def extract_hwpss(aman, signal=None, hwp_angle=None,
+                  bin_signal=True, bins=3600,
+                  lin_reg=True, modes=[1, 2, 3, 4, 6, 8],
+                  apply_prefilt=True, prefilt_cutoff=1.0,
+                  mask_flags=True,
+                  add_stats_to_aman=True, hwpss_stats_name='hwpss_stats',
+                  add_extracted_to_aman=True, hwpss_extract_name='hwpss_extract'):
+    """
+    Extracts HWP synchronous signal (HWPSS) from a time-ordered data (TOD) using linear regression or curve-fitting.
+
+    Parameters
+    ----------
+    aman : AxisManager object
+        The TOD to extract HWPSS from.
+    signal : array-like, optional
+        The TOD signal to use. If not provided, `aman.signal` will be used.
+    hwp_angle : array-like, optional
+        The HWP angle for each sample in `aman`. If not provided, `aman.hwp_angle` will be used.
+    bin_signal : bool, optional
+        Whether to bin the TOD signal into HWP angle bins before extracting HWPSS. Default is `True`.
+    bins : int, optional
+        The number of HWP angle bins to use if `bin_signal` is `True`. Default is 3600.
+    lin_reg : bool, optional
+        Whether to use linear regression to extract HWPSS from the binned signal. If `False`, curve-fitting will be used instead.
+        Default is `True`.
+    modes : list of int, optional
+        The HWPSS harmonic modes to extract. Default is [1, 2, 3, 4, 6, 8].
+    apply_prefilt : bool, optional
+        Whether to apply a high-pass filter to signal before extracting HWPSS. Default is `True`.
+    prefilt_cutoff : float, optional
+        The cutoff frequency of the high-pass filter, in Hz. Only used if `apply_prefilt` is `True`. Default is 1.0.
+    mask_flags : bool, optional
+        Whether to mask out flagged samples before extracting HWPSS. Default is `True`.
+    add_stats_to_aman : bool, optional
```

Comment:
Two thoughts: 

1. Please change axis/axes to field/fields in the documentation here, these are being added as fields and not axes. 

2. Please change `add_stats_to_aman` and `add_extract_to_aman` to `merge_stats` and `merge_extract`. `merge` is the keyword we're using for this in the rest of the tod_ops / flags code so it would be good to keep it similar.